pipeline/alignments/align.py

```diff
@@ -12,101 +19,132 @@
 """
 
 import argparse
+import multiprocessing
 import os
+import shutil
 import subprocess
 import sys
 from contextlib import ExitStack
-from typing import Optional
+from enum import Enum
+from typing import Dict, Optional
 
-import eflomal
 import zstandard
+from tqdm import tqdm
 
+from pipeline.alignments.tokenizer import tokenize
 from pipeline.common.logging import get_logger
 
 logger = get_logger("alignments")
 COMPRESSION_CMD = "zstdmt"
 
 
+class Tokenization(Enum):
+    spaces = "spaces"
+    moses = "moses"
+
+
 def run(
     corpus_src: str,
     corpus_trg: str,
     output_path: str,
     priors_input_path: Optional[str],
     priors_output_path: Optional[str],
+    tokenization: Tokenization,
 ):
     bin = os.environ["BIN"]
+    src = os.environ["SRC"]
+    trg = os.environ["TRG"]
 
     tmp_dir = os.path.join(os.path.dirname(output_path), "tmp")
     os.makedirs(tmp_dir, exist_ok=True)
 
-    if corpus_src.endswith(".zst"):
-        logger.info("Decompressing source corpus...")
-        subprocess.check_call([COMPRESSION_CMD, "-d", "-f", "--rm", corpus_src])
-        corpus_src = corpus_src[:-4]
+    corpus_src = decompress(corpus_src)
+    corpus_trg = decompress(corpus_trg)
 
-    if corpus_trg.endswith(".zst"):
-        logger.info("Decompressing target corpus...")
-        subprocess.check_call([COMPRESSION_CMD, "-d", "-f", "--rm", corpus_trg])
-        corpus_trg = corpus_trg[:-4]
+    if tokenization == Tokenization.moses:
+        tokenized_src, tokenized_trg = corpus_src + ".moses", corpus_trg + ".moses"
+        output_aln = os.path.join(tmp_dir, "aln")
+        # C++ tokenizer can process 100k sentences per second on a single core,
+        # so the chunks to parallelize things should be large enough
+        tokenize(corpus_src, tokenized_src, src, chunk_size=500000)
```

Comment:
Tokenization currently takes ~20 min on 400M sentences dataset and 64 cores machine and doesn't allocate a lot of memory. Even if it's 10GB, it's the eflomal aligner that allocates ~70% on a 512GB machine, so since we're optimizing this all together having large chunks to process it faster makes sense. Also, I experimented with different values to maximize the throughput, and 500K looks fine. I see it tokenizes 1M sentences/sec for English.

Reply:
Sounds good, thanks for the explanation.